### PR TITLE
Improve support for Torch and Jax with `dynamic_one_shot`

### PR DESCRIPTION
**Pull request by mudit2812**
**Context:**
Opened in favour of #5630. Bug fix for #5442. This PR updates `dynamic_one_shot` so that it has better compatibility with the `torch` and `jax` interfaces.

**Description of the Change:**
* Change casting method from `array.astype()` to `qml.math.cast` in the `apply_operation` dispatch for `MidMeasureMP`.
* Update usage of `qml.math` in `dynamic_one_shot`.
* When using `qml.counts`, cast results to ints before converting to strings for lists of MCM values and floats for single MCM values. This is needed because jax arrays are not hashable, and the hash of torch tensors seems to be independent of the value(s) stored inside it. Thus, neither can be used as keys for dictionaries.

**Benefits:**
Better interface support with `dynamic_one_shot`.

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

File: doc/releases/changelog-dev.md [hunk 1]
```diff
@@ -91,6 +91,7 @@
 This release contains contributions from (in alphabetical order):
 
 Pietropaolo Frisoni,
+Soran Jahangiri,
```
Comment (mudit2812) [inline doc/releases/changelog-dev.md:94]: Rebase artifact

File: pennylane/fermi/conversion.py [hunk 2]
```diff
@@ -556,7 +556,7 @@ def _(fermi_operator: FermiSentence, n, ps=False, wire_map=None, tol=None):
     qubit_operator = PauliSentence()  # Empty PS as 0 operator to add Pws to
 
     for fw, coeff in fermi_operator.items():
-        fermi_word_as_ps = parity_transform(fw, n, ps=True)
+        fermi_word_as_ps = bravyi_kitaev(fw, n, ps=True)
```
Comment (mudit2812) [inline pennylane/fermi/conversion.py:559]: Rebase artifact
```suggestion
```

File: tests/fermi/test_bravyi_kitaev.py [hunk 3]
```diff
@@ -995,6 +996,22 @@ def test_fermi_sentence_identity():
...
+                PauliWord({0: "X", 1: "Z"}): 0.5,
+                PauliWord({0: "Y"}): 0.5j,
+                PauliWord({0: "Y", 1: "Z"}): -0.5j,
+                PauliWord({1: "Z", 2: "X", 3: "Z"}): -0.25j,
+                PauliWord({1: "Z", 2: "Y", 3: "Z"}): 0.25,
+                PauliWord({2: "X"}): 0.25j,
+                PauliWord({2: "Y"}): -0.25,
+            }
+        ),
+    ),
```
Comment (mudit2812) [inline tests/fermi/test_bravyi_kitaev.py:1014]: Rebase artifact
```suggestion
```

File: tests/fermi/test_bravyi_kitaev.py [hunk 4]
```diff
@@ -928,6 +928,7 @@ def test_bravyi_kitaev_for_null_operator_fermi_word_ps(operator):
 fw2 = FermiWord({(0, 0): "+", (1, 0): "-"})
 fw3 = FermiWord({(0, 0): "+", (1, 3): "-", (2, 0): "+", (3, 4): "-"})
 fw4 = FermiWord({})
+fw5 = FermiWord({(0, 3): "+", (1, 2): "-"})
```
Comment (mudit2812) [inline tests/fermi/test_bravyi_kitaev.py:931]: Rebase artifact
```suggestion
```

File: doc/releases/changelog-dev.md (hunk 1 above)
Comment (mudit2812) [inline doc/releases/changelog-dev.md:94]: Rebase artifact
```suggestion
```

**Comment by trbromley** [comment]
Should 59938 be the relevant SC story here?

File: pennylane/devices/qubit/apply_operation.py [hunk 5]
```diff
@@ -12,7 +12,7 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 """Functions to apply an operation to a state vector."""
-# pylint: disable=unused-argument
+# pylint: disable=unused-argument, too-many-arguments
```
Comment (trbromley) [inline pennylane/devices/qubit/apply_operation.py:15]: Do we need this at a module level or could we disable just for the function?

File: pennylane/devices/qubit/apply_operation.py [hunk 6]
```diff
@@ -273,31 +295,50 @@ def apply_mid_measure(
...
+            return binomial(prng_key, n, p).astype(int)
+
+    else:
+        binomial_fn = np.random.binomial if rng is None else rng.binomial
+    sample = binomial_fn(1, 1 - prob0)
+    mid_measurements[op] = sample
+
+    # Using apply_operation(qml.QubitUnitary,...) instead of apply_operation(qml.Projector([sample], wire),...)
+    # to select the sample branch enables jax.jit and prevents it from using Python callbacks
+    matrix = qml.math.array([[(sample + 1) % 2, 0.0], [0.0, (sample) % 2]], like=interface)
```
Comment (trbromley) [inline pennylane/devices/qubit/apply_operation.py:320]: This is not a unitary, right? But we're able to get away with this because there's no validation of the matrix?

**Comment by mudit2812** [comment]
[sc-59938]

File: pennylane/devices/qubit/apply_operation.py (hunk 5 above)
Comment (mudit2812) [reply pennylane/devices/qubit/apply_operation.py]: That may be a merge artifact.

File: pennylane/devices/qubit/apply_operation.py (hunk 6 above)
Comment (mudit2812) [reply pennylane/devices/qubit/apply_operation.py]: Another merge artifact. This is the matrix for a projector to apply to the state for postselection.

File: tests/devices/default_qubit/test_default_qubit_native_mcm.py [hunk 7]
```diff
@@ -750,3 +750,85 @@ def func(x):
...
+        m0 = qml.measure(0)
+        qml.RX(0.5 * x, 1)
+        m1 = qml.measure(1, postselect=postselect)
+        qml.cond((m0 + m1) == 2, qml.RY)(2.0 * x, 0)
+        m2 = qml.measure(0)
+
+        mid_measure = 0.5 * m2 if meas_obj == "composite_mcm" else [m1, m2]
+        measurement_key = "wires" if isinstance(meas_obj, list) else "op"
+        measurement_value = mid_measure if isinstance(meas_obj, str) else meas_obj
+        return measure_f(**{measurement_key: measurement_value})
```
[the remark this answers is not in the document]
Comment (mudit2812) [reply tests/devices/default_qubit/test_default_qubit_native_mcm.py:785]: I'm okay with leaving the other tests as they are, since it is much easier to read and understand the tests in their current state. If we start parametrizing more, I'm concerned about making the tests harder to parse.

File: tests/devices/default_qubit/test_default_qubit_native_mcm.py [hunk 8]
```diff
@@ -750,3 +750,86 @@ def func(x):
...
+    """Test that native MCM circuits are executed correctly with Torch"""
+    if measure_f in (qml.var, qml.expval) and (
+        isinstance(meas_obj, list) or meas_obj == "mcm_list"
+    ):
+        pytest.skip("Can't use wires/mcm lists with var or expval")
+
+    import torch
+
+    dev = get_device(shots=shots)
+    param = torch.tensor(np.pi / 3, dtype=torch.float64)
```
Comment (mudit2812) [inline tests/devices/default_qubit/test_default_qubit_native_mcm.py:771]: Using double precision here because using single precision causes a `Probabilities do not sum to 1` error in the step that runs the circuit with the `defer_measurements` transform.

**Review by vincentmr — APPROVED** [review]
Looks good to me, thanks @mudit2812 .

File: tests/devices/default_qubit/test_default_qubit_native_mcm.py [hunk 9]
```diff
@@ -750,3 +750,86 @@ def func(x):
...
+        m0 = qml.measure(0)
+        qml.RX(0.5 * x, 1)
+        m1 = qml.measure(1, postselect=postselect)
+        qml.cond((m0 + m1) == 2, qml.RY)(2.0 * x, 0)
+        m2 = qml.measure(0)
+
+        mid_measure = 0.5 * m2 if meas_obj == "composite_mcm" else [m1, m2]
+        measurement_key = "wires" if isinstance(meas_obj, list) else "op"
+        measurement_value = mid_measure if isinstance(meas_obj, str) else meas_obj
+        return measure_f(**{measurement_key: measurement_value})
```
Comment (dwierichs) [inline tests/devices/default_qubit/test_default_qubit_native_mcm.py:785]: Got to iterate the feedback, this is a nice testing logic :sunglasses:

**Review by dwierichs — CHANGES_REQUESTED** [review]
Thanks @mudit2812 , this is a neat extension of MCM support! :) 

Had three small suggestions, a question for understanding and would like to iterate Vincent's kudo!

File: pennylane/transforms/dynamic_one_shot.py [hunk 10]
```diff
@@ -341,14 +349,13 @@ def gather_mcm(measurement, samples, is_valid):
             counts = qml.math.array(counts, like=interface)
             return counts / qml.math.sum(counts)
         if isinstance(measurement, CountsMP):
-            mcm_samples = [{"".join(str(v) for v in tuple(s)): 1} for s in mcm_samples]
+            mcm_samples = [{"".join(str(int(v)) for v in tuple(s)): 1} for s in mcm_samples]
         return gather_non_mcm(measurement, mcm_samples, is_valid)
+    mcm_samples = qml.math.ravel(qml.math.array(mv.concretize(samples), like=interface))
```
[the remark this answers is not in the document]
Comment (mudit2812) [reply pennylane/transforms/dynamic_one_shot.py:354]: >So the additional leading axis of size 1 in the non-ProbabilityMP case is not required, because it is ravelled away anyways, right?

Yes exactly.

> Was this previously an alias for at_least_1d?

I guess implicitly, yes, but I hadn't really thought about it that way 😅

**Comment by mudit2812** [comment]
I will keep iterating with changing the number of shots/seed until the tests pass, but otherwise this should be good to go.

File: pennylane/transforms/dynamic_one_shot.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (dwierichs) [reply pennylane/transforms/dynamic_one_shot.py:357]: Thanks! :)

**Review by dwierichs — APPROVED** [review]
Thanks @mudit2812 !
